cpp/src/text/subword/load_merges_file.cu

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <text/subword/bpe_tokenizer.cuh>
+
+#include <nvtext/bpe_tokenize.hpp>
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+namespace nvtext {
+namespace detail {
+
+namespace {
+
+struct make_pair_function {
+  /**
+   * @brief Hash the merge pair entry
+   */
+  __device__ cuco::pair_type<hash_value_type, cudf::size_type> operator()(cudf::size_type idx)
+  {
+    auto const result = _hasher(d_strings.element<cudf::string_view>(idx));
+    return cuco::make_pair(result, idx);
+  }
+
+  string_hasher_type const _hasher;
+  cudf::column_device_view const d_strings;
+};
+
+/**
+ * @brief Loads a text file of merge-pairs into a strings column.
+ *
+ * The line position in the file indicates the pair's rank.
+ *
+ * @code{.pseudo}
+ * Format of the file:
+ * #version ..
+ * a1 a2
+ * b1 b2
+ * c1 c2
+ * ...
+ * @endcode
+ *
+ * @param filename_merges Path to text file containing merge-pairs.
+ * @return object containing table elements for the BPE function
+ */
+std::unique_ptr<cudf::column> load_file_to_column(std::string const& filename_merges,
+                                                  rmm::cuda_stream_view stream,
+                                                  rmm::mr::device_memory_resource* mr)
+{
+  std::ifstream merges_file(filename_merges);
+  CUDF_EXPECTS(merges_file.good(), "Could not open " + filename_merges);
+
+  std::vector<char> chars{};
+  std::vector<cudf::offset_type> offsets(1, 0);
+
+  std::string line;
+  std::getline(merges_file, line);
+  std::string version = "#version";
+  if (line.substr(0, version.size()).compare(version) == 0) { std::getline(merges_file, line); }
+
+  // This is a text file delimited only by CR/LF.
+  // TODO: Look into using the CSV reader to load the strings column instead.
```

Comment:
Agreed, it would be good if this code path could avoid file parsing and call an existing API instead.